File: Sources/GKPageScrollView/GKPageDefine.m
//
//  GKPageDefine.m
//  GKPageScrollViewObjc
//
//  Created by gaokun on 2020/10/22.
//  Copyright © 2020 gaokun. All rights reserved.
//

#import "GKPageDefine.h"

@implementation GKPageDefine

+ (BOOL)gk_isNotchedScreen {
    if (@available(iOS 11.0, *)) {
        UIWindow *keyWindow = [GKPageDefine getKeyWindow];
        if (keyWindow) {
            return keyWindow.safeAreaInsets.bottom > 0;
        }
    }
    // 当iOS11以下或获取不到keyWindow时用以下方案
    CGSize screenSize = UIScreen.mainScreen.bounds.size;
    return (CGSizeEqualToSize(screenSize, CGSizeMake(375, 812)) ||
            CGSizeEqualToSize(screenSize, CGSizeMake(812, 375)) ||
            CGSizeEqualToSize(screenSize, CGSizeMake(414, 896)) ||
            CGSizeEqualToSize(screenSize, CGSizeMake(896, 414)) ||
            CGSizeEqualToSize(screenSize, CGSizeMake(390, 844)) ||
            CGSizeEqualToSize(screenSize, CGSizeMake(844, 390)) ||
            CGSizeEqualToSize(screenSize, CGSizeMake(428, 926)) ||
            CGSizeEqualToSize(screenSize, CGSizeMake(926, 428)));
}

+ (CGRect)gk_statusBarFrame {
    CGRect statusBarFrame = CGRectZero;
    if (@available(iOS 13.0, *)) {
        statusBarFrame = [GKPageDefine getKeyWindow].windowScene.statusBarManager.statusBarFrame;
    }
    
    if (CGRectEqualToRect(statusBarFrame, CGRectZero)) {
        statusBarFrame = [UIApplication sharedApplication].statusBarFrame;
    }
    
    if (CGRectEqualToRect(statusBarFrame, CGRectZero)) {
        CGFloat statusBarH = [GKPageDefine gk_isNotchedScreen] ? 44 : 20;
        statusBarFrame = CGRectMake(0, 0, UIScreen.mainScreen.bounds.size.width, statusBarH);
    }
    
    return statusBarFrame;
}

+ (UIWindow *)getKeyWindow {
    UIWindow *window = nil;
    if (@available(iOS 13.0, *)) {
        for (UIWindowScene* windowScene in [UIApplication sharedApplication].connectedScenes) {
            if (windowScene.activationState == UISceneActivationStateForegroundActive) {
                for (UIWindow *w in windowScene.windows) {
                    if (window.isKeyWindow) {
                        window = w;
                        break;
                    }
                }
            }
        }
    }
    
    if (!window) {
        window = [UIApplication sharedApplication].windows.firstObject;
        if (!window.isKeyWindow) {
            UIWindow *keyWindow = [UIApplication sharedApplication].keyWindow;
            if (CGRectEqualToRect(keyWindow.bounds, UIScreen.mainScreen.bounds)) {
                window = keyWindow;
            }
        }
    }
    
    return window;
}

@end

File: Sources/GKPageScrollViewSwift/GKPageDefine.swift
//
//  GKPageDefine.swift
//  GKPageScrollViewSwift
//
//  Created by QuintGao on 2020/5/10.
//  Copyright © 2020 QuintGao. All rights reserved.
//

import UIKit

// 是否是iPhone X系列
public let GKPage_IS_iPhoneX: Bool = GKPageDefine.gk_isNotchedScreen()

// 状态栏高度
public let GKPage_StatusBar_Height: CGFloat = GKPageDefine.gk_statusBarFrame().size.height

// 导航栏+状态栏高度
public let GKPage_NavBar_Height: CGFloat = (GKPage_StatusBar_Height + 44.0)

// 屏幕宽高
public let GKPage_Screen_Width = UIScreen.main.bounds.size.width
public let GKPage_Screen_Height = UIScreen.main.bounds.size.height

open class GKPageDefine {
    
    
//    public static let isNotchedScreen: Bool = {
//        if #available(iOS 11.0, *) {
//            var window = keyWindow()
//            if window == nil {
//                // keyWindow还没有创建时，通过创建临时window获取安全区域
//                window = UIWindow(frame: UIScreen.main.bounds)
//                if window!.safeAreaInsets.bottom <= 0 {
//                    let viewController = UIViewController()
//                    window?.rootViewController = viewController
//                }
//            }
//
//            if window!.safeAreaInsets.bottom > 0 {
//                return true
//            }
//        } else {
//            return false
//        }
//        return false
//    }()
    
    open class func gk_isNotchedScreen() -> Bool {
        if #available(iOS 11.0, *) {
            var window = getKeyWindow()
            if window == nil {
                // keyWindow还没有创建，通过创建临时window获取安全区域
                window = UIWindow(frame: UIScreen.main.bounds)
                if window!.safeAreaInsets.bottom <= 0 {
                    let viewController = UIViewController()
                    window?.rootViewController = viewController
                }
            }
            
            if window!.safeAreaInsets.bottom > 0 {
                return true
            }
        }
        return false
    }
    
    open class func gk_statusBarFrame() -> CGRect {
        var statusBarFrame = CGRect.zero
        if #available(iOS 13.0, *) {
            statusBarFrame = GKPageDefine.getKeyWindow()?.windowScene?.statusBarManager?.statusBarFrame ?? .zero
        }
        
        if statusBarFrame == .zero {
            statusBarFrame = UIApplication.shared.statusBarFrame
        }
        
        if statusBarFrame == .zero {
            let statusBarH: CGFloat = GKPageDefine.gk_isNotchedScreen() ? 44.0 : 20.0
            statusBarFrame = CGRect(x: 0, y: 0, width: UIScreen.main.bounds.size.width, height: statusBarH)
        }
        return statusBarFrame
    }
    
    fileprivate class func getKeyWindow() -> UIWindow? {
        var window: UIWindow?
        if #available(iOS 13.0, *) {
            window = UIApplication.shared.connectedScenes
                .filter({$0.activationState == .foregroundActive})
                .map({$0 as? UIWindowScene})
                .compactMap({$0})
                .first?.windows
                .filter({$0.isKeyWindow}).first
        }
        
        if window == nil {
            window = UIApplication.shared.windows.first { $0.isKeyWindow }
            if window == nil {
                window = UIApplication.shared.keyWindow
            }
        }
        return window
    }
}
